core/src/main/scala/org/bitcoins/core/wallet/fee/FeeUnit.scala

```diff
@@ -15,24 +15,53 @@ sealed abstract class FeeUnit {
   final def *(long: Long): CurrencyUnit = Satoshis(toLong * long / scaleFactor)
   def *(tx: Transaction): CurrencyUnit = calc(tx)
 
+  def factory: FeeUnitFactory[FeeUnit]
+
   /** The coefficient the denominator in the unit is multiplied by,
     * for example sats/kilobyte -> 1000
     */
-  def scaleFactor: Long
+  def scaleFactor: Long = factory.scaleFactor
 
   require(scaleFactor > 0,
           s"Scale factor cannot be less than or equal to 0, got $scaleFactor")
 
   /** Takes the given transaction returns a size that will be used for calculating the fee rate.
     * This is generally the denominator in the unit, ie sats/byte
     */
-  def txSizeForCalc(tx: Transaction): Long
+  def txSizeForCalc(tx: Transaction): Long = factory.txSizeForCalc(tx)
 
   /** Calculates the fee for the transaction using this fee rate, rounds down satoshis */
   final def calc(tx: Transaction): CurrencyUnit = this * txSizeForCalc(tx)
   def toLong: Long = currencyUnit.satoshis.toLong
 }
 
+trait FeeUnitFactory[+T <: FeeUnit] {
+
+  /** The coefficient the denominator in the unit is multiplied by,
+    * for example sats/kilobyte -> 1000
+    */
+  def scaleFactor: Long
+
+  require(scaleFactor > 0,
+          s"Scale factor cannot be less than or equal to 0, got $scaleFactor")
+
+  /** Takes the given transaction returns a size that will be used for calculating the fee rate.
+    * This is generally the denominator in the unit, ie sats/byte
+    */
+  def txSizeForCalc(tx: Transaction): Long
+
+  /** Creates an instance T where the value given is the numerator of the fee unit
+    */
+  def fromLong(long: Long): T
+
+  def calc(inputAmount: CurrencyUnit, tx: Transaction): T = {
```

Comment:
What exactly is `inputAmount` here? A scala doc would be nice. You don't have an index if it's a specific input. If it's for all inputs we probably should rename to `totalInputAmount`